pkg/reconciler/apibinding/apibinding_reconcile_test.go

```diff
@@ -0,0 +1,764 @@
+/*
+Copyright 2022 The KCP Authors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package apibinding
+
+import (
+	"context"
+	"errors"
+	"testing"
+
+	"github.com/stretchr/testify/require"
+
+	corev1 "k8s.io/api/core/v1"
+	apiextensionsv1 "k8s.io/apiextensions-apiserver/pkg/apis/apiextensions/v1"
+	apierrors "k8s.io/apimachinery/pkg/api/errors"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/runtime"
+	"k8s.io/apimachinery/pkg/runtime/schema"
+	"k8s.io/apimachinery/pkg/types"
+	"k8s.io/utils/pointer"
+
+	apisv1alpha1 "github.com/kcp-dev/kcp/pkg/apis/apis/v1alpha1"
+	conditionsv1alpha1 "github.com/kcp-dev/kcp/third_party/conditions/apis/conditions/v1alpha1"
+	"github.com/kcp-dev/kcp/third_party/conditions/util/conditions"
+)
+
+type wantCondition struct {
+	Type     conditionsv1alpha1.ConditionType
+	Status   corev1.ConditionStatus
+	Reason   string
+	Severity conditionsv1alpha1.ConditionSeverity
+}
+
+func requireConditionMatches(t *testing.T, g conditions.Getter, w wantCondition) {
+	c := conditions.Get(g, w.Type)
+	require.NotNil(t, c, "missing condition %q", w.Type)
+	require.Equal(t, w.Status, c.Status)
+	require.Equal(t, w.Severity, c.Severity)
+	require.Equal(t, w.Reason, c.Reason)
+}
+
+func TestPhaseReconciler(t *testing.T) {
+	tests := map[string]struct {
+		apiBinding          *apisv1alpha1.APIBinding
+		apiExport           *apisv1alpha1.APIExport
+		getAPIExportError   error
+		apiResourceSchemas  map[string]*apisv1alpha1.APIResourceSchema
+		wantReconcileStatus reconcileStatus
+		wantPhase           apisv1alpha1.APIBindingPhaseType
+		wantCondition       *wantCondition
+		wantError           bool
+	}{
+		"empty phase becomes binding": {
+			apiBinding:          &apisv1alpha1.APIBinding{},
+			wantPhase:           apisv1alpha1.APIBindingPhaseBinding,
+			wantReconcileStatus: reconcileStatusContinue,
+		},
+		"bound becomes binding when export changes": {
+			apiBinding: new(bindingBuilder).
+				WithClusterName("org:ws").
+				WithName("my-binding").
+				WithWorkspaceReference("new-workspace", "new-export").
+				WithPhase(apisv1alpha1.APIBindingPhaseBound).
+				WithBoundAPIExport("some-workspace", "some-export").
+				Build(),
+			wantPhase:           apisv1alpha1.APIBindingPhaseBinding,
```

Comment:
(No action required) If there only 2 phases, maybe simplify the test case definition by using `wantBound: bool` and set the explicit phase in the test?